src/SectorView.cpp

```diff
@@ -47,6 +47,128 @@ enum DetailSelection {
+{
+	return std::max(std::max(box[XMAX] - box[XMIN], box[YMAX] - box[YMIN]), box[ZMAX] - box[ZMIN]);
+}
+
+// helper function
+// cache - source of sector and system data
+// crd - coordinates of the origin point in the sector, l.y.
+// originSector - coordinates of the sector containing the origin point
+// targetSector - coordinates of the sector in which we are looking for the nearest system
+// min_dist - here we will accumulate the minimum distance
+// best_sys - here we will accumulate the nearest system
+static void try_sector_for_nearer_system(cache_t cache, const vector3f &crd, const SystemPath &originSector, const SystemPath &targetSector, float &min_dist, SystemPath &best_sys)
+{
+	RefCountedPtr<Sector> ps = cache->GetCached(targetSector);
+	vector3f base(
+		targetSector.sectorX - originSector.sectorX,
+		targetSector.sectorY - originSector.sectorY,
+		targetSector.sectorZ - originSector.sectorZ);
+	base *= Sector::SIZE;
+	if (ps->m_systems.size()) {
+		for (unsigned int i = 0; i < ps->m_systems.size(); i++) {
+			Sector::System *ss = &ps->m_systems[i];
+			vector3f dx = crd - (ss->GetPosition() + base);
+			float dist = dx.LengthSqr();
+			if (dist < min_dist) {
+				min_dist = dist;
+				best_sys = ss->GetPath();
+			}
+		}
+	}
+}
+
+// the function searches for the closest star to the given coordinates
+// cache - source of sector and system data
+// pos -  absolute coordinates of a point (in sectors, float)
+// returns the found SystemPath, (system closest to the given position)
+// if returned !SystemPath.IsSystemPath(), it means that nothing was found
+static SystemPath nearest_system_to_pos(cache_t cache, const vector3f &pos)
+{
+	// maximum size in sectors of the search box
+	const int MAX_SEARCH_DIST = 20;
+	// extract the coordinates of the sector, and the coordinates of the position within the sector
+	SystemPath sectorPos(floor(pos.x), floor(pos.y), floor(pos.z));
+	vector3f crd(FFRAC(pos.x), FFRAC(pos.y), FFRAC(pos.z));
+	crd *= Sector::SIZE; // convert to l.y.
+	// directions of "growth" of the search box, if necessary
+	box_t enlarge = { -1, 1, -1, 1, -1, 1 };
+	// remember the initial sector coordinates and point coordinates as boxes with size 0, for ease of comparison
+	box_t originBox = {
+		sectorPos.sectorX, sectorPos.sectorX,
+		sectorPos.sectorY, sectorPos.sectorY,
+		sectorPos.sectorZ, sectorPos.sectorZ
+	};
+	std::array<float, SIDES> posBox = { crd.x, crd.x, crd.y, crd.y, crd.z, crd.z };
+	// we will remember in which box we were already looked, so as not to repeat the search
+	box_t prevSearchBox = originBox;
+	// here we will accumulate the minimum distance
+	float min_dist = FLT_MAX;
+	// here we will accumulate the nearest system (this will be the result of the function)
+	SystemPath best_sys;
+	// the first iteration of the search - only in the sector where the base point is located
+	box_t searchBox = originBox;
+	++searchBox[XMAX];
+	++searchBox[YMAX];
+	++searchBox[ZMAX];
```

Comment:
This section in particular catches my eye as the first thing that would need improvement; instead of manually incrementing the maximum values of the box, I'd recommend changing the for-loop condition to `<=` instead; not only does it cause significant confusion about whether a box with `min = (1,1,1); max = (2,2,2)` has size==1 or size==2, it also generally is contradictory to existing AABB logic in the codebase which uses bounds tests of `p >= min && p <= max`.

Reply:
@Web-eWorks This is because otherwise the bottom and top edges will have different semantics (?), I mean you have to write different comparison functions for the min and max edges. That is, if you follow your advice, a box of size 1 will be `min (3,3,3) max (3,3,3)`, the point with absolute coordinates `(3.5 3.5 3.5)` will be inside this box, although its coordinates will be greater than the maximum values of the box, therefore, comparing with the upper bound, you will need to add 1.

I don't understand what the logic of our AABB has to do with it, because there are real numbers, and for them `>` and `>=` are almost the same.

Reply:
Hmm... I see your point now that I understand what your actual bounds test does. In any case, I'd recommend constructing the max values of the search box directly as `x + 1, y + 1` etc. instead of doing it in this manner.

Reply:
Done!